---
layout: page
element: lecture
title: "Case study: Using nanopore sequencing to study DNA replication dynamics"
---

We are providing a brief summary of the lectures presented live during the training course.
For the hands-on sessions, all materials are provided in the respective web pages.

### Conclusions

BrdU incorporation into the genome and subsequent detection is organism agnostic

There are two ways to study replication dynamics

These methods can be used to identify origins, fork direction, replication termination sites, and pauses


### References

#### DNAscent in *Saccharomyces cerevisiae*:

Müller CA, Boemo MA, Spingardi P, Kessler BM, Kriaucionis S, Simpson JT, Nieduszynski CA. Capturing the dynamics of genome replication on individual ultra-long nanopore sequence reads. Nat Methods. 2019 May;16(5):429-436. doi: 10.1038/s41592-019-0394-y. Epub 2019 Apr 22. PMID: 31011185.

Boemo MA. DNAscent v2: detecting replication forks in nanopore sequencing data with deep learning. BMC Genomics. 2021 Jun 9;22(1):430. doi: 10.1186/s12864-021-07736-6. PMID: 34107894; PMCID: PMC8191041.

#### DNAscent (EdU-BrdU pulse chase) in human:

Mathew J.K. Jones, Subash Kumar Rai, Pauline L. Pfuderer, Alexis Bonfim-Melo, Julia K. Pagan, Paul R. Clarke, Sarah E. McClelland, Michael A. Boemo. A high-resolution, nanopore-based artificial intelligence assay for DNA replication stress in human cancer cells.
bioRxiv 2022.09.22.509021; doi: https://doi.org/10.1101/2022.09.22.509021

#### DNAscent in *Plasmodium falciparum*:

Totañes FIG, Gockel J, Chapman SE, Bártfai R, Boemo MA, Merrick CJ. A genome-wide map of DNA replication at single-molecule resolution in the malaria parasite Plasmodium falciparum. Nucleic Acids Res. 2023 Apr 11;51(6):2709-2724. doi: 10.1093/nar/gkad093. PMID: 36808528; PMCID: PMC10085703.

#### DNAscent in *Giardia lamblia*:

da Silva MS, Vitarelli MO, Viala VL, Tsantarlis K, da Silva Pires D, Franco TA, de Azevedo ILMJ, Elias MC, Tonelli RR. Clues on the dynamics of DNA replication in Giardia lamblia. J Cell Sci. 2023 May 15;136(10):jcs260828. doi: 10.1242/jcs.260828. Epub 2023 May 30. PMID: 37248991.



#### FORK-seq in *Saccharomyces cerevisiae*:

Theulot B, Lacroix L, Arbona JM, Millot GA, Jean E, Cruaud C, Pellet J, Proux F, Hennion M, Engelen S, Lemainque A, Audit B, Hyrien O, Le Tallec B. Genome-wide mapping of individual replication fork velocities using nanopore sequencing. Nat Commun. 2022 Jun 8;13(1):3295. doi: 10.1038/s41467-022-31012-0. PMID: 35676270; PMCID: PMC9177527.

Hennion M, Arbona JM, Lacroix L, Cruaud C, Theulot B, Tallec BL, Proux F, Wu X, Novikova E, Engelen S, Lemainque A, Audit B, Hyrien O. FORK-seq: replication landscape of the Saccharomyces cerevisiae genome by nanopore sequencing. Genome Biol. 2020 May 26;21(1):125. doi: 10.1186/s13059-020-02013-3. PMID: 32456659; PMCID: PMC7251829.

Hennion M, Theulot B, Arbona JM, Audit B, Hyrien O. FORK-seq: Single-Molecule Profiling of DNA Replication. Methods Mol Biol. 2022;2477:107-128. doi: 10.1007/978-1-0716-2257-5_8. PMID: 35524115.

#### FORK-seq in *Plasmodium falciparum*:

Castellano CM, Lacroix L, Mathis E, Prorok P, Hennion M, Lopez-Rubio JJ, Méchali M, Gomes AR. The genetic landscape of origins of replication in P. falciparum. Nucleic Acids Res. 2023 Dec 1:gkad1103. doi: 10.1093/nar/gkad1103. Epub ahead of print. PMID: 38038269.
